hermes/bin/src/wasm/module.rs

```diff
@@ -0,0 +1,140 @@
+//! WASM module implementation.
+//! Wrapper over the `wasmtime::Module`, `wasmtime::Instance` etc. structs which
+//! define a WASM module abstraction with capability to interact with it.
+//!
+//! All implementation based on [wasmtime](https://crates.io/crates/wasmtime) crate dependency.
+
+use std::error::Error;
+
+use wasmtime::{
+    Config as WasmConfig, Engine as WasmEngine, InstancePre as WasmModuleInstance,
+    Linker as WasmLinker, Module as WasmModule, Store as WasmStore, WasmParams, WasmResults,
+};
+
+use super::context::Context;
+
+/// Interface for linking WASM imports
+pub(crate) trait LinkImport<Context> {
+    /// Link imports to the `wasmtime::Linker`
+    fn link(&self, linker: &mut WasmLinker<Context>) -> Result<(), Box<dyn Error>>;
+}
+
+/// Structure defines an abstraction over the WASM module
+/// It instantiates the module with the provided context data,
+/// links all provided imports to the module instance,
+/// handles an internal state of the WASM module.
+///
+/// The primary goal for it is to make a WASM state *immutable* along WASM module
+/// execution. It means that `Module::call_func` execution does not have as side effect
+/// for the WASM module's state, it becomes unchanged.
+pub(crate) struct Module {
+    /// `wasmtime::InstancePre` entity
+    ///
+    /// A reason why it is used a `wasmtime::InstancePre` instead of `wasmtime::Instance`
+    /// partially described in this [RFC](https://github.com/bytecodealliance/rfcs/blob/main/accepted/shared-host-functions.md).
+    /// It separates and optimizes the linkage of the imports to the WASM runtime from the
+    /// module actual initialization process.
+    instance: WasmModuleInstance<Context>,
+
+    /// `wasmtime::Engine` entity
+    engine: WasmEngine,
+
+    /// `Context` entity
+    context: Context,
+}
+
+impl Module {
+    /// Instantiate WASM module
+    ///
+    /// # Errors
+    ///  - `wasmtime::Error`: WASM call error
+    #[allow(dead_code)]
+    pub(crate) fn new(
+        context: Context, module_bytes: &[u8], imports: &[Box<dyn LinkImport<Context>>],
+    ) -> Result<Self, Box<dyn Error>> {
+        let mut config = WasmConfig::new();
+        config.wasm_component_model(true);
+        config.consume_fuel(false);
+
+        let engine = WasmEngine::new(&config)?;
```

Comment:
There should only be one engine, for all modules.
You can make an `engine` crate which creates a new engine, and then we just keep re-using it.
Pass the engine into the `new` as a parameter.
Still save it in the Module struct, just in case in future we do something like have a different engine for each application.